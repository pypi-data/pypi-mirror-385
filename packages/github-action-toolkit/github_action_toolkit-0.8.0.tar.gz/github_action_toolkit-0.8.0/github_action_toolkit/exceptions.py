"""
Custom exception taxonomy for github-action-toolkit.

Provides specific exception classes for better error handling and debugging.
"""

from __future__ import annotations


class GitHubActionError(Exception):
    """Base exception for all github-action-toolkit errors."""


class EnvironmentVariableError(GitHubActionError):
    """
    Raised when required environment variables are missing or invalid.

    This typically occurs when the toolkit is not running in a GitHub Actions
    context or required environment variables are not set.
    """


class InputError(GitHubActionError):
    """
    Raised when user input is invalid or cannot be parsed.

    This occurs when converting input values to specific types fails or when
    required inputs are missing.
    """


class GitOperationError(GitHubActionError):
    """
    Raised when git operations fail.

    This occurs during repository operations like clone, checkout, commit, push, etc.
    """


class GitHubAPIError(GitHubActionError):
    """
    Raised when GitHub API operations fail.

    This occurs when interacting with GitHub APIs for artifacts, PRs, etc.
    """


class ConfigurationError(GitHubActionError):
    """
    Raised when configuration is invalid or incomplete.

    This occurs when required configuration parameters are missing or invalid.
    """


class CacheNotFoundError(GitHubActionError):
    """Raised when a cache entry is not found."""


class CacheRestoreError(GitHubActionError):
    """Raised when cache restoration fails."""


class CacheSaveError(GitHubActionError):
    """Raised when cache save fails."""


class CancellationRequested(GitHubActionError):
    """
    Raised when a cancellation signal (SIGTERM, SIGINT) is received.

    This allows code to handle cancellation gracefully by catching this exception.
    """


class RateLimitError(GitHubActionError):
    """Raised when GitHub API rate limit is exceeded."""

    def __init__(self, reset_time: int | None = None):
        self.reset_time: int | None = reset_time
        super().__init__("GitHub API rate limit exceeded")


class APIError(GitHubActionError):
    """Raised when GitHub API returns an error."""

    def __init__(self, status_code: int, message: str):
        self.status_code: int = status_code
        super().__init__(f"GitHub API error ({status_code}): {message}")
