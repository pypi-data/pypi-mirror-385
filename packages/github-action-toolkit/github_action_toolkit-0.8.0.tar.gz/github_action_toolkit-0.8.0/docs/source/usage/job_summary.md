# Job Summary

Create rich, formatted summaries for GitHub Actions workflows using a fluent builder API. Job summaries appear on the workflow run summary page and support tables, code blocks, images, collapsible sections, and more.

## Table of Contents

- [Fluent Builder API](#fluent-builder-api)
- [Basic Elements](#basic-elements)
- [Tables](#tables)
- [Code Blocks](#code-blocks)
- [Images](#images)
- [Collapsible Sections](#collapsible-sections)
- [Template API](#template-api)
- [Legacy Functions](#legacy-functions)
- [Size Limits and Security](#size-limits-and-security)

## Fluent Builder API

The `JobSummary` class provides a fluent API for building rich summaries. All methods return `self` for easy chaining.

### Basic Usage

```python
from github_action_toolkit import JobSummary

summary = JobSummary()
summary.add_heading("Test Results", 1)
summary.add_separator()
summary.add_list(["Test 1: ✓ Passed", "Test 2: ✓ Passed", "Test 3: ✗ Failed"])
summary.write()
```

### Fluent Chaining

Chain multiple methods for concise code:

```python
summary = JobSummary()
(summary
    .add_heading("CI Pipeline Report", 1)
    .add_separator()
    .add_quote("All tests passed!")
    .add_break()
    .add_link("View Dashboard", "https://example.com")
    .write()
)
```

## Basic Elements

### Headings

Add headings at levels 1-6:

```python
summary.add_heading("Main Title", 1)
summary.add_heading("Subtitle", 2)
summary.add_heading("Section", 3)
```

### Text and Line Breaks

Add raw text, line breaks, and separators:

```python
summary.add_raw("Plain text content")
summary.add_eol()  # Add newline
summary.add_break()  # Add <br>
summary.add_separator()  # Add <hr>
```

### Quotes

Add blockquotes with optional citations:

```python
summary.add_quote("This is a quote")
summary.add_quote("Quote with citation", cite="Author Name")
```

### Links

Add hyperlinks:

```python
summary.add_link("GitHub", "https://github.com")
```

### Lists

Create ordered or unordered lists:

```python
# Unordered list
summary.add_list(["Item 1", "Item 2", "Item 3"])

# Ordered list
summary.add_list(["First", "Second", "Third"], ordered=True)
```

## Tables

Create tables with headers and rows. The first row is automatically treated as headers.

### Simple Tables

```python
rows = [
    ["Name", "Status", "Duration"],
    ["test_auth", "✓ Pass", "0.5s"],
    ["test_db", "✓ Pass", "1.2s"],
    ["test_api", "✗ Fail", "0.8s"],
]
summary.add_table(rows)
```

### Advanced Tables with Cell Options

Use dictionaries for cells that need special attributes:

```python
rows = [
    [{"data": "Header 1", "header": True}, {"data": "Header 2", "header": True}],
    [{"data": "Spanning Cell", "colspan": "2"}],
    ["Regular", "Cells"],
]
summary.add_table(rows)
```

Cell dictionary keys:
- `data`: Cell content (required)
- `header`: True to render as `<th>` instead of `<td>`
- `colspan`: Number of columns to span
- `rowspan`: Number of rows to span

## Code Blocks

Add code blocks with optional syntax highlighting:

```python
# Without language
summary.add_code_block("print('hello world')")

# With language
summary.add_code_block("""
def greet(name):
    return f"Hello, {name}!"
""", "python")
```

## Images

Add images with alt text and optional dimensions:

```python
# Basic image
summary.add_image("https://example.com/image.png", "Alt text")

# With dimensions
summary.add_image(
    "https://example.com/chart.png",
    "Performance Chart",
    width="400",
    height="300"
)
```

## Collapsible Sections

Create collapsible details sections:

```python
summary.add_details(
    "Click to expand",
    "Hidden content that appears when expanded"
)
```

## Template API

The `JobSummaryTemplate` class provides pre-built templates for common use cases.

### Test Report

```python
from github_action_toolkit import JobSummaryTemplate

summary = JobSummaryTemplate.test_report(
    title="Unit Test Results",
    passed=45,
    failed=3,
    skipped=2,
    duration="12.5s"
)
summary.write()
```

### Coverage Report

```python
coverage_data = {
    "core/auth.py": 98.5,
    "core/database.py": 92.3,
    "api/handlers.py": 87.6,
}

summary = JobSummaryTemplate.coverage_report(
    "Code Coverage",
    coverage_data
)
summary.write()
```

### Deployment Report

```python
summary = JobSummaryTemplate.deployment_report(
    title="Production Deployment",
    environment="production",
    status="success",
    version="v2.5.0",
    url="https://app.example.com"
)
summary.write()
```

### Benchmark Report

```python
benchmarks = {
    "API Response Time": {
        "Average": "125ms",
        "P95": "250ms",
        "P99": "500ms",
    },
    "Database Query": {
        "Average": "45ms",
        "P95": "80ms",
    },
}

summary = JobSummaryTemplate.benchmark_report(
    "Performance Benchmarks",
    benchmarks
)
summary.write()
```

### Combining Templates

Templates return `JobSummary` objects, so you can add custom content:

```python
summary = JobSummaryTemplate.test_report(
    title="CI Results",
    passed=50,
    failed=0,
    duration="15.2s"
)

# Add custom content
summary.add_separator()
summary.add_heading("Additional Notes", 2)
summary.add_quote("All quality gates passed!")
summary.write()
```

## Legacy Functions

These functions provide backward compatibility with earlier versions.

### `append_job_summary(markdown_text)`

Append markdown text to the job summary:

```python
from github_action_toolkit import append_job_summary

append_job_summary("# Test Summary")
append_job_summary("- Point 1")
```

### `overwrite_job_summary(markdown_text)`

Replace the entire job summary:

```python
from github_action_toolkit import overwrite_job_summary

overwrite_job_summary("# New Summary")
```

### `remove_job_summary()`

Remove the job summary file:

```python
from github_action_toolkit import remove_job_summary

remove_job_summary()
```

## Size Limits and Security

### Size Limits

Job summaries are limited to 1 MiB (1,048,576 bytes). The API automatically enforces this limit:

```python
summary = JobSummary()
summary.add_raw("..." * 1000000)  # Large content
summary.write()  # Raises ValueError if too large
```

### Content Sanitization

All content is automatically sanitized to prevent XSS attacks. HTML special characters are escaped:

```python
summary = JobSummary()
summary.add_heading("<script>alert('xss')</script>")
# Rendered as: &lt;script&gt;alert('xss')&lt;/script&gt;
```

### Buffer Management

Control when content is written:

```python
summary = JobSummary()
summary.add_heading("Title", 1)
summary.add_raw("Content")

# Check if empty
if not summary.is_empty():
    # Get as string without writing
    content = summary.stringify()
    
    # Clear buffer
    summary.clear()
    
    # Write to file (buffer is cleared after write)
    summary.write()
    
    # Overwrite file
    summary.write(overwrite=True)
```

## Complete Example

Here's a complete example combining multiple features:

```python
from github_action_toolkit import JobSummary, JobSummaryTemplate

# Use template for test results
summary = JobSummaryTemplate.test_report(
    title="CI Pipeline Results",
    passed=42,
    failed=0,
    duration="8.5s"
)

# Add coverage information
summary.add_separator()
summary.add_heading("Code Coverage", 2)
summary.add_table([
    ["Module", "Coverage"],
    ["Core", "95%"],
    ["API", "92%"],
    ["Utils", "87%"],
])

# Add deployment info
summary.add_separator()
summary.add_heading("Deployment", 2)
summary.add_quote("✓ Successfully deployed to staging")
summary.add_link("View Application", "https://staging.example.com")

# Add collapsible logs
summary.add_break()
summary.add_details("View Build Logs", """
Building...
Running tests...
All tests passed!
Deploying...
Deployment successful!
""")

# Write the summary
summary.write()
```

## Reference

For more information, see the [GitHub Actions documentation on job summaries](https://docs.github.com/en/actions/using-workflows/workflow-commands-for-github-actions#adding-a-job-summary).
