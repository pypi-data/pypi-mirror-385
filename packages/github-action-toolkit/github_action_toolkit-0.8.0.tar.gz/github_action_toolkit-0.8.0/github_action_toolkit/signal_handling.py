"""
Signal handling utilities for graceful cancellation.

Provides support for handling SIGTERM and other signals for graceful shutdown
in GitHub Actions workflows.
"""

from __future__ import annotations

import signal
from collections.abc import Callable
from typing import Any

from .exceptions import CancellationRequested
from .print_messages import warning


class CancellationHandler:
    """
    Unified handler for managing cancellation signals in GitHub Actions workflows.

    Provides methods to enable/disable signal handling, register cleanup handlers,
    and check cancellation status.
    """

    def __init__(self) -> None:
        self._handlers: list[Callable[[], None]] = []
        self._original_handlers: dict[signal.Signals, Any] = {}
        self._enabled: bool = False

    def register(self, handler: Callable[[], None]) -> None:
        """
        Register a handler to be called on cancellation.

        The handler will be called when SIGTERM or SIGINT is received, before
        raising CancellationRequested. Handlers should perform cleanup operations.

        Example:
            def cleanup():
                print("Cleaning up resources...")

            cancellation.register(cleanup)

        :param handler: Function to call on cancellation (takes no arguments)
        """
        self._handlers.append(handler)

    def enable(self) -> None:
        """
        Enable automatic handling of cancellation signals.

        Sets up signal handlers for SIGTERM and SIGINT that will:
        1. Call all registered cancellation handlers
        2. Raise CancellationRequested exception

        This allows code to gracefully handle cancellation in GitHub Actions workflows.

        Example:
            cancellation = CancellationHandler()
            cancellation.enable()
            try:
                # Your long-running operation
                pass
            except CancellationRequested:
                print("Operation was cancelled")
        """
        if self._enabled:
            return

        self._original_handlers[signal.SIGTERM] = signal.signal(signal.SIGTERM, self._handle_signal)
        self._original_handlers[signal.SIGINT] = signal.signal(signal.SIGINT, self._handle_signal)

        self._enabled = True

    def disable(self) -> None:
        """
        Disable automatic handling of cancellation signals.

        Restores original signal handlers. Useful for testing or when you need
        to temporarily disable cancellation handling.
        """
        if not self._enabled:
            return

        for sig, original_handler in self._original_handlers.items():
            if original_handler is not None:
                signal.signal(sig, original_handler)

        self._original_handlers.clear()
        self._enabled = False

    def is_enabled(self) -> bool:
        """
        Check if cancellation support is currently enabled.

        :returns: True if cancellation handlers are registered, False otherwise
        """
        return self._enabled

    def _handle_signal(self, signum: int, frame: Any) -> None:  # pyright: ignore[reportUnusedParameter]
        """Handle cancellation signals by calling registered handlers and raising exception."""
        signal_name = signal.Signals(signum).name
        warning(f"Received {signal_name} signal. Initiating graceful shutdown...")

        for handler in self._handlers:
            try:
                handler()
            except Exception as e:  # noqa: BLE001
                warning(f"Error in cancellation handler: {e}")

        raise CancellationRequested(f"Operation cancelled by {signal_name} signal")
