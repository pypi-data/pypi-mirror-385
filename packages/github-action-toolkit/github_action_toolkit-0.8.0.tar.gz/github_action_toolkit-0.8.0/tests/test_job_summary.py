# pyright: reportPrivateUsage=false
# pyright: reportUnusedVariable=false
# pyright: reportUnusedParameter=false
# pyright: reportMissingParameterType=false
# pyright: reportUnknownVariableType=false
# pyright: reportUnknownParameterType=false
# pyright: reportUnknownMemberType=false

import os
from typing import Any
from unittest import mock

import pytest

import github_action_toolkit as gat
import github_action_toolkit.job_summary as gat_job_summary


@pytest.mark.parametrize(
    "test_input,expected",
    [
        ("# test", "# test"),
        ("# test%0A", "# test\n"),
        ("- %25test", "- %test"),
        ("**%0Dtest**", "**\rtest**"),
    ],
)
def test__clean_markdown_string(test_input: str, expected: str) -> None:
    assert gat_job_summary._clean_markdown_string(test_input) == expected


def test_append_job_summary(tmpdir: Any) -> None:
    file = tmpdir.join("summary")

    with mock.patch.dict(os.environ, {"GITHUB_STEP_SUMMARY": file.strpath}):
        gat.append_job_summary("# TEST")
        gat.append_job_summary("- point 1")

    assert file.read() == "# TEST\n- point 1\n"


def test_overwrite_job_summary(tmpdir: Any) -> None:
    file = tmpdir.join("summary")

    with mock.patch.dict(os.environ, {"GITHUB_STEP_SUMMARY": file.strpath}):
        gat.append_job_summary("# TEST")
        gat.overwrite_job_summary("- point 1")

    assert file.read() == "- point 1\n"


def test_remove_job_summary(tmpdir: Any) -> None:
    file = tmpdir.join("summary")

    with mock.patch.dict(os.environ, {"GITHUB_STEP_SUMMARY": file.strpath}):
        gat.remove_job_summary()

    assert os.path.isfile(file.strpath) is False


# JobSummary builder tests


def test_job_summary_basic_usage() -> None:
    summary = gat.JobSummary()
    assert summary.is_empty()

    summary.add_raw("Hello World")
    assert not summary.is_empty()
    assert summary.stringify() == "Hello World"

    summary.clear()
    assert summary.is_empty()


def test_job_summary_add_heading() -> None:
    summary = gat.JobSummary()
    summary.add_heading("Test Heading", 1)
    assert "<h1>Test Heading</h1>" in summary.stringify()

    summary.clear()
    summary.add_heading("Subheading", 2)
    assert "<h2>Subheading</h2>" in summary.stringify()

    summary.clear()
    summary.add_heading("Test", 10)  # Invalid level, should default to 1
    assert "<h1>Test</h1>" in summary.stringify()


def test_job_summary_add_separator() -> None:
    summary = gat.JobSummary()
    summary.add_separator()
    assert "<hr>" in summary.stringify()


def test_job_summary_add_break() -> None:
    summary = gat.JobSummary()
    summary.add_break()
    assert "<br>" in summary.stringify()


def test_job_summary_add_quote() -> None:
    summary = gat.JobSummary()
    summary.add_quote("This is a quote")
    assert "<blockquote>This is a quote</blockquote>" in summary.stringify()

    summary.clear()
    summary.add_quote("Quote with citation", "Author")
    result = summary.stringify()
    assert "<blockquote" in result
    assert 'cite="Author"' in result
    assert "Quote with citation" in result


def test_job_summary_add_link() -> None:
    summary = gat.JobSummary()
    summary.add_link("GitHub", "https://github.com")
    result = summary.stringify()
    assert '<a href="https://github.com">GitHub</a>' in result


def test_job_summary_add_code_block() -> None:
    summary = gat.JobSummary()
    summary.add_code_block("print('hello')")
    assert "<pre><code>print('hello')</code></pre>" in summary.stringify()

    summary.clear()
    summary.add_code_block("def test(): pass", "python")
    result = summary.stringify()
    assert '<pre lang="python">' in result
    assert "def test(): pass" in result


def test_job_summary_add_list() -> None:
    summary = gat.JobSummary()
    summary.add_list(["Item 1", "Item 2", "Item 3"])
    result = summary.stringify()
    assert "<ul>" in result
    assert "<li>Item 1</li>" in result
    assert "<li>Item 2</li>" in result
    assert "<li>Item 3</li>" in result
    assert "</ul>" in result

    summary.clear()
    summary.add_list(["First", "Second"], ordered=True)
    result = summary.stringify()
    assert "<ol>" in result
    assert "<li>First</li>" in result
    assert "</ol>" in result


def test_job_summary_add_table_simple() -> None:
    summary = gat.JobSummary()
    rows = [
        ["Header 1", "Header 2"],
        ["Row 1 Col 1", "Row 1 Col 2"],
        ["Row 2 Col 1", "Row 2 Col 2"],
    ]
    summary.add_table(rows)
    result = summary.stringify()
    assert "<table>" in result
    assert "<th>Header 1</th>" in result
    assert "<th>Header 2</th>" in result
    assert "<td>Row 1 Col 1</td>" in result
    assert "</table>" in result


def test_job_summary_add_table_with_dict_cells() -> None:
    summary = gat.JobSummary()
    rows: list[list[str | dict[str, str | bool | int]]] = [
        [{"data": "Header 1", "header": True}, {"data": "Header 2", "header": True}],
        [{"data": "Spanning", "colspan": "2"}],
        ["Regular", "Cell"],
    ]
    summary.add_table(rows)
    result = summary.stringify()
    assert "<th>Header 1</th>" in result
    assert "<th>Header 2</th>" in result
    assert 'colspan="2"' in result
    assert "<td>Regular</td>" in result


def test_job_summary_add_details() -> None:
    summary = gat.JobSummary()
    summary.add_details("Click to expand", "Hidden content here")
    result = summary.stringify()
    assert "<details>" in result
    assert "<summary>Click to expand</summary>" in result
    assert "Hidden content here" in result
    assert "</details>" in result


def test_job_summary_add_image() -> None:
    summary = gat.JobSummary()
    summary.add_image("https://example.com/image.png", "Test Image")
    result = summary.stringify()
    assert '<img src="https://example.com/image.png" alt="Test Image">' in result

    summary.clear()
    summary.add_image("image.png", "Alt text", width="100", height="50")
    result = summary.stringify()
    assert 'width="100"' in result
    assert 'height="50"' in result


def test_job_summary_sanitization() -> None:
    summary = gat.JobSummary()
    summary.add_heading("<script>alert('xss')</script>")
    result = summary.stringify()
    assert "&lt;script&gt;" in result
    assert "<script>" not in result


def test_job_summary_fluent_chaining() -> None:
    summary = gat.JobSummary()
    result = (
        summary.add_heading("Test Report", 1)
        .add_separator()
        .add_list(["Test 1: Pass", "Test 2: Pass"])
        .add_break()
        .add_quote("All tests passed!")
        .stringify()
    )
    assert "<h1>Test Report</h1>" in result
    assert "<hr>" in result
    assert "<ul>" in result
    assert "<blockquote>" in result


def test_job_summary_write(tmpdir: Any) -> None:
    file = tmpdir.join("summary")

    with mock.patch.dict(os.environ, {"GITHUB_STEP_SUMMARY": file.strpath}):
        summary = gat.JobSummary()
        summary.add_heading("Test Summary", 1).add_raw("Some content").write()

        assert "<h1>Test Summary</h1>" in file.read()
        assert summary.is_empty()  # Buffer should be cleared after write


def test_job_summary_write_overwrite(tmpdir: Any) -> None:
    file = tmpdir.join("summary")
    file.write("Old content")

    with mock.patch.dict(os.environ, {"GITHUB_STEP_SUMMARY": file.strpath}):
        summary = gat.JobSummary()
        summary.add_heading("New Summary", 1).write(overwrite=True)

        content = file.read()
        assert "<h1>New Summary</h1>" in content
        assert "Old content" not in content


def test_job_summary_write_append(tmpdir: Any) -> None:
    file = tmpdir.join("summary")
    file.write("Existing content\n")

    with mock.patch.dict(os.environ, {"GITHUB_STEP_SUMMARY": file.strpath}):
        summary = gat.JobSummary()
        summary.add_heading("Appended Summary", 1).write(overwrite=False)

        content = file.read()
        assert "Existing content" in content
        assert "<h1>Appended Summary</h1>" in content


def test_job_summary_write_no_env_var() -> None:
    with mock.patch.dict(os.environ, {}, clear=True):
        summary = gat.JobSummary()
        summary.add_heading("Test", 1)

        with pytest.raises(ValueError, match="GITHUB_STEP_SUMMARY"):
            summary.write()


def test_job_summary_size_limit(tmpdir: Any) -> None:
    file = tmpdir.join("summary")

    with mock.patch.dict(os.environ, {"GITHUB_STEP_SUMMARY": file.strpath}):
        summary = gat.JobSummary()
        large_content = "x" * (gat_job_summary.MAX_SUMMARY_SIZE + 1)
        summary.add_raw(large_content)

        with pytest.raises(ValueError, match="Summary exceeds maximum size"):
            summary.write()


# JobSummaryTemplate tests


def test_template_test_report() -> None:
    summary = gat.JobSummaryTemplate.test_report(
        title="Unit Tests", passed=10, failed=2, skipped=1, duration="5.2s"
    )
    result = summary.stringify()
    assert "<h1>Unit Tests</h1>" in result
    assert "2 test(s) failed" in result
    assert "<td>Total Tests</td>" in result
    assert "<td>13</td>" in result
    assert "<td>10</td>" in result
    assert "<td>2</td>" in result
    assert "<td>1</td>" in result
    assert "5.2s" in result


def test_template_test_report_all_passed() -> None:
    summary = gat.JobSummaryTemplate.test_report(title="Unit Tests", passed=10, failed=0)
    result = summary.stringify()
    assert "All tests passed" in result
    assert "<td>10</td>" in result


def test_template_coverage_report() -> None:
    modules = {"core": 95.5, "utils": 87.2, "api": 92.0}
    summary = gat.JobSummaryTemplate.coverage_report("Code Coverage", modules)
    result = summary.stringify()
    assert "<h1>Code Coverage</h1>" in result
    assert "Good coverage" in result
    assert "91.6%" in result  # Average
    assert "core" in result
    assert "95.5%" in result


def test_template_coverage_report_empty() -> None:
    summary = gat.JobSummaryTemplate.coverage_report("Code Coverage", {})
    result = summary.stringify()
    assert "No coverage data available" in result


def test_template_deployment_report() -> None:
    summary = gat.JobSummaryTemplate.deployment_report(
        title="Deployment",
        environment="production",
        status="success",
        version="1.2.3",
        url="https://example.com",
    )
    result = summary.stringify()
    assert "<h1>Deployment</h1>" in result
    assert "production" in result
    assert "success" in result
    assert "1.2.3" in result
    assert "https://example.com" in result


def test_template_benchmark_report() -> None:
    benchmarks = {
        "API Response Time": {"Average": "120ms", "P95": "250ms", "P99": "500ms"},
        "Database Query": {"Average": "45ms", "P95": "80ms"},
    }
    summary = gat.JobSummaryTemplate.benchmark_report("Performance Benchmarks", benchmarks)
    result = summary.stringify()
    assert "<h1>Performance Benchmarks</h1>" in result
    assert "API Response Time" in result
    assert "120ms" in result
    assert "Database Query" in result
    assert "45ms" in result


def test_template_benchmark_report_empty() -> None:
    summary = gat.JobSummaryTemplate.benchmark_report("Performance Benchmarks", {})
    result = summary.stringify()
    assert "No benchmark data available" in result


# Snapshot tests for job summary formatting


def test_job_summary_formatting(snapshot: Any, tmpdir: Any) -> None:
    """Snapshot test for job summary formatting."""
    file = tmpdir.join("summary")

    with mock.patch.dict(os.environ, {"GITHUB_STEP_SUMMARY": file.strpath}):
        gat.append_job_summary("# Test Summary")
        gat.append_job_summary("## Section 1")
        gat.append_job_summary("- Item 1")
        gat.append_job_summary("- Item 2")
        gat.append_job_summary("")
        gat.append_job_summary("## Section 2")
        gat.append_job_summary("Some **bold** text and *italic* text.")

    result = file.read()
    assert result == snapshot


def test_job_summary_with_special_chars(snapshot: Any, tmpdir: Any) -> None:
    """Snapshot test for job summary with special characters."""
    file = tmpdir.join("summary")

    with mock.patch.dict(os.environ, {"GITHUB_STEP_SUMMARY": file.strpath}):
        gat.append_job_summary("# Test with %25 percent")
        gat.append_job_summary("Line with %0A newline")
        gat.append_job_summary("Line with %0D carriage return")

    result = file.read()
    assert result == snapshot


def test_job_summary_overwrite_snapshot(snapshot: Any, tmpdir: Any) -> None:
    """Snapshot test for job summary overwrite."""
    file = tmpdir.join("summary")

    with mock.patch.dict(os.environ, {"GITHUB_STEP_SUMMARY": file.strpath}):
        gat.append_job_summary("# Initial content")
        gat.append_job_summary("- Point 1")
        gat.overwrite_job_summary("# New content")
        gat.append_job_summary("- New point")

    result = file.read()
    assert result == snapshot


def test_job_summary_complex_markdown(snapshot: Any, tmpdir: Any) -> None:
    """Snapshot test for complex markdown formatting."""
    file = tmpdir.join("summary")

    complex_markdown = """
# GitHub Action Summary

## Build Results
- ✅ Build passed
- ✅ Tests passed
- ⚠️  Coverage at 85%

## Details
| Metric | Value |
|--------|-------|
| Tests  | 104   |
| Lines  | 397   |

### Code Quality
All checks passed successfully!
"""

    with mock.patch.dict(os.environ, {"GITHUB_STEP_SUMMARY": file.strpath}):
        gat.append_job_summary(complex_markdown)

    result = file.read()
    assert result == snapshot


# Additional snapshot tests for JobSummary builder


def test_job_summary_builder_table_snapshot(snapshot: Any) -> None:
    """Snapshot test for JobSummary table building."""
    summary = gat.JobSummary()
    rows = [
        ["Name", "Status", "Duration"],
        ["Build", "✅ Success", "2m 15s"],
        ["Test", "✅ Success", "5m 42s"],
        ["Deploy", "⚠️  Warning", "1m 30s"],
    ]
    summary.add_heading("CI/CD Pipeline", 2).add_table(rows)
    result = summary.stringify()
    assert result == snapshot


def test_job_summary_builder_code_blocks_snapshot(snapshot: Any) -> None:
    """Snapshot test for JobSummary code blocks."""
    summary = gat.JobSummary()
    summary.add_heading("Code Examples", 2).add_code_block(
        'def hello():\n    print("Hello, World!")', "python"
    ).add_break().add_code_block("npm install github-action-toolkit", "bash")
    result = summary.stringify()
    assert result == snapshot


def test_job_summary_builder_mixed_content_snapshot(snapshot: Any) -> None:
    """Snapshot test for JobSummary with mixed content types."""
    summary = gat.JobSummary()
    (
        summary.add_heading("Test Report", 1)
        .add_separator()
        .add_quote("Testing is an integral part of software development.", "Anonymous")
        .add_break()
        .add_list(
            ["Unit tests: 150 passed", "Integration tests: 45 passed", "E2E tests: 20 passed"]
        )
        .add_separator()
        .add_link("View detailed report", "https://example.com/report")
        .add_details("Click to expand logs", "Log content here\nLine 2\nLine 3")
    )
    result = summary.stringify()
    assert result == snapshot


def test_job_summary_template_test_report_snapshot(snapshot: Any) -> None:
    """Snapshot test for test report template."""
    summary = gat.JobSummaryTemplate.test_report(
        title="Integration Test Results",
        passed=85,
        failed=3,
        skipped=2,
        duration="12m 34s",
    )
    result = summary.stringify()
    assert result == snapshot


def test_job_summary_template_coverage_report_snapshot(snapshot: Any) -> None:
    """Snapshot test for coverage report template."""
    modules = {
        "authentication": 92.5,
        "database": 88.3,
        "api": 95.1,
        "utils": 78.9,
        "models": 91.0,
    }
    summary = gat.JobSummaryTemplate.coverage_report("Test Coverage Report", modules)
    result = summary.stringify()
    assert result == snapshot


def test_job_summary_sanitization_snapshot(snapshot: Any) -> None:
    """Snapshot test for HTML sanitization in job summary."""
    summary = gat.JobSummary()
    summary.add_heading("<script>alert('xss')</script> Malicious Title").add_raw(
        "<img src=x onerror=alert('xss')>"
    ).add_quote("<b>Bold</b> and <i>italic</i> tags")
    result = summary.stringify()
    assert result == snapshot
