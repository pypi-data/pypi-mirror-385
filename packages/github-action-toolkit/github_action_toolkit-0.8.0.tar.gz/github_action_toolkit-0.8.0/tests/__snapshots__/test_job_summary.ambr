# serializer version: 1
# name: test_job_summary_builder_code_blocks_snapshot
  '''
  <h2>Code Examples</h2>
  <pre lang="python"><code>def hello():
      print("Hello, World!")</code></pre>
  <br>
  <pre lang="bash"><code>npm install github-action-toolkit</code></pre>
  
  '''
# ---
# name: test_job_summary_builder_mixed_content_snapshot
  '''
  <h1>Test Report</h1>
  <hr>
  <blockquote cite="Anonymous">Testing is an integral part of software development.</blockquote>
  <br>
  <ul>
  <li>Unit tests: 150 passed</li>
  <li>Integration tests: 45 passed</li>
  <li>E2E tests: 20 passed</li>
  </ul>
  <hr>
  <a href="https://example.com/report">View detailed report</a><details>
  <summary>Click to expand logs</summary>
  Log content here
  Line 2
  Line 3
  </details>
  
  '''
# ---
# name: test_job_summary_builder_table_snapshot
  '''
  <h2>CI/CD Pipeline</h2>
  <table>
  <tr>
  <th>Name</th>
  <th>Status</th>
  <th>Duration</th>
  </tr>
  <tr>
  <td>Build</td>
  <td>✅ Success</td>
  <td>2m 15s</td>
  </tr>
  <tr>
  <td>Test</td>
  <td>✅ Success</td>
  <td>5m 42s</td>
  </tr>
  <tr>
  <td>Deploy</td>
  <td>⚠️  Warning</td>
  <td>1m 30s</td>
  </tr>
  </table>
  
  '''
# ---
# name: test_job_summary_complex_markdown
  '''
  
  # GitHub Action Summary
  
  ## Build Results
  - ✅ Build passed
  - ✅ Tests passed
  - ⚠️  Coverage at 85%
  
  ## Details
  | Metric | Value |
  |--------|-------|
  | Tests  | 104   |
  | Lines  | 397   |
  
  ### Code Quality
  All checks passed successfully!
  
  
  '''
# ---
# name: test_job_summary_formatting
  '''
  # Test Summary
  ## Section 1
  - Item 1
  - Item 2
  
  ## Section 2
  Some **bold** text and *italic* text.
  
  '''
# ---
# name: test_job_summary_overwrite_snapshot
  '''
  # New content
  - New point
  
  '''
# ---
# name: test_job_summary_sanitization_snapshot
  '''
  <h1>&lt;script&gt;alert('xss')&lt;/script&gt; Malicious Title</h1>
  <img src=x onerror=alert('xss')><blockquote>&lt;b&gt;Bold&lt;/b&gt; and &lt;i&gt;italic&lt;/i&gt; tags</blockquote>
  
  '''
# ---
# name: test_job_summary_template_coverage_report_snapshot
  '''
  <h1>Test Coverage Report</h1>
  <hr>
  <blockquote>✓ Good coverage: 89.2% average</blockquote>
  <table>
  <tr>
  <th>Module</th>
  <th>Coverage</th>
  </tr>
  <tr>
  <td>api</td>
  <td>✓ 95.1%</td>
  </tr>
  <tr>
  <td>authentication</td>
  <td>✓ 92.5%</td>
  </tr>
  <tr>
  <td>database</td>
  <td>✓ 88.3%</td>
  </tr>
  <tr>
  <td>models</td>
  <td>✓ 91.0%</td>
  </tr>
  <tr>
  <td>utils</td>
  <td>⚠ 78.9%</td>
  </tr>
  </table>
  
  '''
# ---
# name: test_job_summary_template_test_report_snapshot
  '''
  <h1>Integration Test Results</h1>
  <hr>
  <blockquote>✗ 3 test(s) failed</blockquote>
  <table>
  <tr>
  <th>Metric</th>
  <th>Count</th>
  </tr>
  <tr>
  <td>Total Tests</td>
  <td>90</td>
  </tr>
  <tr>
  <td>✓ Passed</td>
  <td>85</td>
  </tr>
  <tr>
  <td>✗ Failed</td>
  <td>3</td>
  </tr>
  <tr>
  <td>⊘ Skipped</td>
  <td>2</td>
  </tr>
  <tr>
  <td>Duration</td>
  <td>12m 34s</td>
  </tr>
  </table>
  
  '''
# ---
# name: test_job_summary_with_special_chars
  '''
  # Test with % percent
  Line with 
   newline
  Line with 
   carriage return
  
  '''
# ---
